File: solutions/2026-08-12/cpp/2026-08-12-length-of-longest-subarray-with-at-most-k-frequency.cpp
```cpp
// Problem: Length of Longest Subarray With at Most K Frequency
// Problem Link: https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency/
//
// Approach:
// This problem can be solved efficiently using the sliding window technique.
// We maintain a window [left, right] and a frequency map to store the count of each
// element within the current window.
// We expand the window by moving the 'right' pointer. For each element added, we
// increment its count in the frequency map.
// If at any point the frequency of an element in the window exceeds 'k', we need
// to shrink the window from the left. We move the 'left' pointer and decrement
// the count of the element at the 'left' index in the frequency map until the
// frequency of all elements is at most 'k'.
// In each step, after ensuring the window is "good" (all element frequencies <= k),
// we update the maximum length of the good subarray found so far.
//
// Time Complexity: O(N), where N is the length of the input array `nums`.
// Both `left` and `right` pointers traverse the array at most once.
// The hash map operations (insertion, deletion, lookup) take O(1) on average.
//
// Space Complexity: O(M), where M is the number of distinct elements in `nums`.
// In the worst case, all elements could be distinct, leading to O(N) space.
// However, the problem constraints on `nums[i]` do not limit the number of distinct elements directly,
// but in practice, it's bound by the number of unique values encountered.

#include <vector>
#include <unordered_map>
#include <algorithm>

class Solution {
public:
    int maxLength(std::vector<int>& nums, int k) {
        // `freq_map` will store the frequency of each element within the current window.
        std::unordered_map<int, int> freq_map;
        
        // `left` is the left boundary of our sliding window.
        int left = 0;
        
        // `max_len` stores the maximum length of a "good" subarray found so far.
        int max_len = 0;
        
        // Iterate through the array with the `right` pointer, expanding the window.
        for (int right = 0; right < nums.size(); ++right) {
            // Increment the frequency of the element at the current `right` pointer.
            freq_map[nums[right]]++;
            
            // If the frequency of the element at `nums[right]` exceeds `k`,
            // we need to shrink the window from the left.
            while (freq_map[nums[right]] > k) {
                // Decrement the frequency of the element at the `left` pointer.
                freq_map[nums[left]]--;
                // If the frequency becomes zero, we can optionally remove it from the map
                // to save space, though it's not strictly necessary for correctness.
                // if (freq_map[nums[left]] == 0) {
                //     freq_map.erase(nums[left]);
                // }
                
                // Move the `left` pointer to the right, effectively shrinking the window.
                left++;
            }
            
            // After ensuring the current window [left, right] is "good" (all frequencies <= k),
            // calculate its length and update `max_len` if this window is longer.
            // The length of the current window is `right - left + 1`.
            max_len = std::max(max_len, right - left + 1);
        }
        
        // Return the maximum length of a good subarray found.
        return max_len;
    }
};
```

File: solutions/2026-08-12/python/2026-08-12-length-of-longest-subarray-with-at-most-k-frequency.py
```python
# Summary: Find the longest contiguous subarray where each element appears at most k times.
# Link: https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency/
# Approach: This problem can be efficiently solved using the sliding window technique.
# We maintain a window [left, right] and a frequency map (dictionary) to store the counts of elements within the current window.
# We expand the window by moving the 'right' pointer. For each element added, we update its count in the frequency map.
# If at any point the frequency of an element exceeds 'k', it means the current window is no longer "good".
# To make the window "good" again, we shrink the window from the left by moving the 'left' pointer.
# As we shrink, we decrement the count of the element leaving the window.
# We continue shrinking until the frequency of all elements in the window is less than or equal to 'k'.
# At each valid step (when the window is "good"), we update the maximum length of the good subarray found so far.
# Time Complexity: O(N), where N is the length of the input array `nums`.
# Each element is visited at most twice (once by the `right` pointer and once by the `left` pointer).
# The operations within the loop (dictionary lookups, updates, comparisons) take O(1) on average.
# Space Complexity: O(M), where M is the number of distinct elements in `nums`.
# In the worst case, if all elements are distinct, the space complexity can be O(N).
# The frequency map stores the counts of elements within the current window.

from collections import defaultdict

class Solution:
    def maxFrequency(self, nums: list[int], k: int) -> int:
        # Initialize the left pointer of the sliding window.
        left = 0
        # Initialize the maximum length of a good subarray found so far.
        max_length = 0
        # Use a defaultdict to store the frequency of elements within the current window.
        # defaultdict(int) automatically initializes the count to 0 for new keys.
        frequency_map = defaultdict(int)

        # Iterate through the array with the right pointer of the sliding window.
        for right in range(len(nums)):
            # Get the element at the current right pointer.
            current_element = nums[right]
            # Increment the frequency of the current element in the map.
            frequency_map[current_element] += 1

            # Check if the frequency of the current element exceeds k.
            # If it does, we need to shrink the window from the left until the condition is met.
            while frequency_map[current_element] > k:
                # Get the element at the current left pointer.
                left_element = nums[left]
                # Decrement the frequency of the element that is about to leave the window.
                frequency_map[left_element] -= 1
                # If the frequency of the left_element becomes 0 after decrementing,
                # it means this element is no longer present in the window.
                # While not strictly necessary for correctness (as frequency_map[left_element] will be 0),
                # it can slightly optimize by not keeping keys with zero counts if memory is a concern,
                # but for this problem, it doesn't significantly impact performance.
                # if frequency_map[left_element] == 0:
                #     del frequency_map[left_element]

                # Move the left pointer one step to the right to shrink the window.
                left += 1

            # At this point, the current window [left, right] is a "good" subarray
            # because the frequency of all elements is at most k.
            # Calculate the length of the current good subarray.
            current_window_length = right - left + 1
            # Update the maximum length found so far.
            max_length = max(max_length, current_window_length)

        # Return the maximum length of the longest good subarray.
        return max_length

```